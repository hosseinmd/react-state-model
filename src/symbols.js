export const IS_MODEL = Symbol("IS_MODEL");
export const IS_STORE = Symbol("IS_STORE");
