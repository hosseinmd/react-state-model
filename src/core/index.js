"use strict";

export { getSnapshot } from "./lib/setState";
export { addListener } from "./lib/Listener";

import { setState } from "./lib/setState";
import { IS_STORE } from "../symbols";

/**
 * @template S
 * @template A
 * @typedef {S & A} Store
 */

/**
 * @template S , A
 * @typedef {(store: Store<S, A>) => S} Initializer
 */

/**
 * @template S , A
 * @param {S} state
 * @param {object} initialState
 * @param {A} [actions]
 * @param {Initializer<S, A>} [initializer]
 * @returns {Store<S, A>}
 */
export const createStore = (state, initialState = {}, actions, initializer) => {
  const store = {
    ...state,
    ...initialState,
    __status: IS_STORE,
    __listeners: {},
    __types: state,
  };
  actions && Object.assign(store, actions(setState.bind(store), store));
  initializer && Object.assign(store, initializer(store));
  return store;
};
