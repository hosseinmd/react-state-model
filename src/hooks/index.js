import { useState, useEffect } from "react";
import { addListener } from "../core";

/**
 * @template S , A
 * @param {import("../core").Store<S, A>} store
 * @returns {(sensitiveStateKeys: (keyof S)[], listener: (S) => void) => [S, A]}
 */
export function createHooks(store) {
  function useHook(sensitiveStateKeys, listener) {
    if (typeof listener !== "function") listener = useState()[1];

    useEffect(() => {
      return addListener(store, listener, sensitiveStateKeys).remove;
    }, []);

    return store;
  }
  return useHook;
}

/**
 * @template S , A
 * @param {S} initialState
 * @param {A} actions
 * @param {(keyof S)[]} sensitiveStateKeys
 * @param {()=>void} listener
 * @returns {[S,A]}
 */
export function useLocalModel(
  model,
  initialState,
  sensitiveStateKeys,
  listener,
) {
  const useStore = useState(() => {
<<<<<<< HEAD
    const localStore = createStore(initialState, actions);
    return createHooks(localStore);
  }, [])[0];
=======
    const localStore = model.create(initialState);
    return createHooks(localStore);
  })[0];
>>>>>>> model
  return useStore(sensitiveStateKeys, listener);
}
