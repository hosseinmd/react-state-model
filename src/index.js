import { createStore, addListener, getSnapshot } from "./core";
import { IS_MODEL } from "./symbols";
export * from "./hooks";


/**
 * @template S
 * @template A
 * @typedef {object} Model
 * @property {() => import("./core").Store<S,A>} create
 * @property {(listener: (state: S) => void, sensitiveStatesKey: (keyof S)[]) => Model<S, A>} addListener
 * @property {(_initializer: import("./core").Initializer<S,A>)=> Model<S,A>} initializer
 */

/**
 * @template S
 * @template A
 * @param {S} state
 * @param {(setState:(S)=>void,self: import("./core").Store<S,{}>) => A } [getActions]
 * @returns {Model<S, A>}
 */
export const model = (state, getActions) => {
  let initializer = undefined;
  const listeners = [];

  const organizer = {
    __status: IS_MODEL,
    initializer(_initializer) {
      initializer = _initializer;
      return organizer;
    },
    create(initialState) {
      const store = createStore(state, initialState, getActions, initializer);
      (listeners || []).forEach(listenerArg => {
        addListener(
          store,
          listenerArg.listener,
          listenerArg.sensitiveStatesKey,
        );
      });

      return store;
    },
    addListener(listener, sensitiveStatesKey) {
      const listenerArg = { listener, sensitiveStatesKey };
      listeners.push(listenerArg);

      return organizer;
    },
  };
  return organizer;
};

export { addListener, getSnapshot };
