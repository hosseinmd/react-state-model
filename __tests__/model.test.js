import { model, addListener, getSnapshot } from "../src";

const Todo = model(
  {
    title: "",
    done: false,
  },
  (setState, self) => ({
    toggle() {
      setState({ done: !self.done });
    },
  }),
)
  .initializer(self => ({}))
  .addListener(snapshot => {
    console.log(snapshot);
  });

const Store = model({
  /**@type {} */
  todos: [Todo],
});

// create an instance from a snapshot
const store = Store.create({
  todos: [
    Todo.create({
      title: "Get coffee",
    }),
  ],
});

test("store ", () => {
  expect(getSnapshot(store)).toMatchSnapshot();

  store.todos[0].toggle();

  expect(getSnapshot(store)).toMatchSnapshot();
});
