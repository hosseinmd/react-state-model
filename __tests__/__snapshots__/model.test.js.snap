// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store  1`] = `
Object {
  "todos": Array [
    Object {
      "done": false,
      "title": "Get coffee",
    },
  ],
}
`;

exports[`store  2`] = `
Object {
  "todos": Array [
    Object {
      "done": true,
      "title": "Get coffee",
    },
  ],
}
`;
